cirq-core/cirq/transformers/routing/mapping_manager.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, TYPE_CHECKING
+import networkx as nx
+
+from cirq import ops, protocols
+
+if TYPE_CHECKING:
+    import cirq
+
+
+class MappingManager:
+    """Class that keeps track of the mapping of logical to physical qubits and provides
+    convenience methods for distance queries on the physical qubits.
+
+    Qubit variables with the characer 'p' preppended to them are physical and qubits with  the
+    character 'l' preppended to them are logical qubits.
+    """
+
+    def __init__(self, device_graph: nx.Graph, initial_mapping: Dict[ops.Qid, ops.Qid]) -> None:
+        """Initializes MappingManager.
+
+        Args:
+            device_graph: connectivity graph of qubits in the hardware device.
+            circuit_graph: connectivity graph of the qubits in the input circuit.
+            initial_mapping: the initial mapping of logical (keys) to physical qubits (values).
+        """
+        self.device_graph = device_graph
+        self._map = initial_mapping.copy()
+        self._inverse_map = {v: k for k, v in self._map.items()}
+        self._induced_subgraph = nx.induced_subgraph(self.device_graph, self._map.values())
+        self._shortest_paths_matrix = dict(nx.all_pairs_shortest_path(self._induced_subgraph))
```

Comment:
This seems (potentially) expensive, what's the size of this matrix? `O(M * N ** 2)`? Do we have any idea of how slow the mapping manager gets for circuits with O(100s) / O(1000s) of qubits? 

We can consider two alternatives:
* Use `@cached_property` on `dist_on_device` and `shortest_path` so we only compute and store the shortest paths for pairs of nodes which are actually queried, instead of storing for all pairs. 
* We can just store a `distance` matrix and a `parent` matrix instead such that `parent[i][j]` stores that parent node when computing shortest path from `i` to `j`. This would bring down the initial preprocessing and storage cost and each query can be answered in O(length of shortest path), by iterating on parents (eg: `while(x != i) x = parent[i][x]`). We can also potentially cache the shortest path queries using `@cached_property` so we answer it in O(1) the next time.

Reply:
I tried this using random graphs of size n=100, m=400 (< 1 second) and size n=1000, m=4000 (~14 seconds) with no assumptions on connectivity or planarity of the graphs. This might be an issue if we plan on routing near 1000 qubits sometime soon, although at that point I doubt the bottleneck would be coming from this step...

Reply:
The runtime for computing the `distance` matrix is still O(n^3) and in practice takes the same amount of time as computing the whole matrix of all pairs shortest paths (13-14 seconds on my machine with n=1000 and m=4000). So I went ahead and implemented the first solution using the cache. My only worry is that if we are querying distances between close to all O(n^2) pairs then with this implementation we end up with O(n^2 * (m +nlogn)) = O(n^3 * logn) time only for querying distance, amortized throughout the execution of the entire algorithm.